substrate/coins/pallet/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+      for (account, coin, amount) in self.accounts.iter() {
+        Coins::<T>::set(account, coin, Some(*amount));
+        let current_supply = Supply::<T>::get(coin);
+        Supply::<T>::set(coin, current_supply.checked_add(*amount).unwrap());
+      }
+    }
+  }
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+      // burn the fees collected previous block
+      let k1: T::AccountId = FEE_ACCOUNT.into();
+      let amount = Self::coins(k1, Coin::Serai).unwrap_or(0);
+      if amount != 0 {
+        // we can unwrap, we are not burning more then what we have.
+        Self::burn_internal(&FEE_ACCOUNT.into(), Coin::Serai, amount, None).unwrap();
+      }
+      Weight::zero() // TODO
+    }
+  }
+
+  impl<T: Config> Pallet<T> {
+    /// Returns the balance of a given account for `coin`.
+    pub fn balance(of: &T::AccountId, coin: Coin) -> SubstrateAmount {
+      Self::coins(of, coin).unwrap_or(0)
+    }
+
+    fn increase_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
+      // update the supply
+      let new_supply = Self::supply(coin).checked_add(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
+      Ok(())
+    }
+
+    fn decrease_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
+      // update the supply
+      let new_supply = Self::supply(coin).checked_sub(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
+      Ok(())
+    }
+
+    /// Mints amount at the given account, errors if amount overflows.
+    pub fn mint(at: &T::AccountId, coin: Coin, amount: SubstrateAmount) -> Result<(), Error<T>> {
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
+
+      // add amount to account
+      let new_amount = Self::coins(at, coin)
+        .unwrap_or(0)
+        .checked_add(amount)
+        .ok_or(Error::<T>::AmountOverflowed)?;
+
+      // save
+      Coins::<T>::set(at, coin, Some(new_amount));
+      Self::increase_supply(coin, amount)?;
+
+      let balance = Balance { coin, amount: Amount(amount) };
+      Self::deposit_event(Event::Mint { address: SeraiAddress(at.0), balance });
+      Ok(())
+    }
+
+    /// Burns amount at the given account, errors if not enough funds to burn.
+    pub fn burn_internal(
+      at: &T::AccountId,
+      coin: Coin,
+      amount: SubstrateAmount,
+      instruction: Option<OutInstruction>,
+    ) -> Result<(), Error<T>> {
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
+
+      // sub amount from account
+      let new_amount =
+        Self::coins(at, coin).unwrap_or(0).checked_sub(amount).ok_or(Error::<T>::NotEnoughFunds)?;
+
+      if new_amount == 0 {
+        Coins::<T>::remove(at, coin);
+      } else {
+        Coins::<T>::set(at, coin, Some(new_amount));
+      }
```

Comment:
decrease_balance_internal to share this with transfer? Same commentary for increase_balance_internal.